homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,378 @@
+            """Trigger."""
+            await self.async_trigger(instance)
+
+        async def stop(self, executed_at: datetime) -> None:
+            """Revert."""
+            await self.async_revert(instance)
+
+        self._async_trigger_listener = async_track_point_in_time(
+            self._hass, start, start_dt
+        )
+
+        if end_dt:
+            self._async_trigger_listener = async_track_point_in_time(
+                self._hass, stop, end_dt
+            )
+
+    @callback
+    def _get_next_instance_datetimes(
+        self,
+    ) -> Tuple[Optional[datetime], Optional[datetime]]:
+        """Get the next starting (and, optionally, ending) datetimes."""
+        if not self._initial_instance_scheduled:
+            self._initial_instance_scheduled = True
+            return (self.start_datetime, self.end_datetime)
+
+        if self.recurrence:
+            start = self.recurrence.after(dt_util.utcnow(), inc=True)
+
+            # The recurrence has reached its end:
+            if not start:
+                return (None, None)
+
+            if self.end_datetime:
+                end = start + self.instance_duration
+            else:
+                end = None
+
+            return (start, end)
+
+        return (None, None)
+
+    def as_dict(self) -> dict:
+        """Return the schedule as a dict."""
+        the_dict = {
+            CONF_ENTITY_ID: self.entity_id,
+            CONF_START_DATETIME: self.start_datetime.isoformat(),
+        }
+
+        if self.end_datetime:
+            the_dict[CONF_END_DATETIME] = self.end_datetime.isoformat()
+        else:
+            the_dict[CONF_END_DATETIME] = None
+
+        if self.recurrence:
+            the_dict[CONF_RECURRENCE] = str(self.recurrence)
+        else:
+            the_dict[CONF_RECURRENCE] = None
+
+        return the_dict
+
+    async def async_revert(self, instance: ScheduleInstance) -> None:
+        """Revert the instance."""
+        await instance.async_revert()
+        await self._async_schedule()
+
+    async def async_trigger(self, instance: ScheduleInstance) -> None:
+        """Trigger the instance."""
+        await instance.async_trigger()
+        if not self.instance_duration:
+            await self._async_schedule()
+
+    @callback
+    def async_turn_off(self) -> None:
+        """Disable the schedule."""
+        if self.active_instance:
+            self.active_instance.async_cancel()
+            self.active_instance = None
+        if self._async_revert_listener:
+            self._async_revert_listener()
+            self._async_revert_listener = None
+        if self._async_trigger_listener:
+            self._async_trigger_listener()
+            self._async_trigger_listener = None
+
+        self._is_on = False
+
+    @callback
+    def async_turn_on(self) -> None:
+        """Enable the schedule."""
+        if self.expired:
+            _LOGGER.error("Refusing to turn on expired schedule: %s", self)
```

Comment:
Let's raise instead of swallowing the error. It should not be called. `raise ValueError("Cannot turn on expired schedule")`